executor/src/lib.rs

```diff
@@ -311,569 +308,256 @@ where
             parent_id, id
         );
 
-        let (resp_sender, resp_receiver) = oneshot::channel();
-        match self
-            .command_sender
-            .lock()
-            .expect("Failed to lock mutex.")
-            .as_ref()
-        {
-            Some(sender) => sender
-                .send(Command::ExecuteBlock {
-                    executable_block: ExecutableBlock {
-                        transactions,
-                        parent_trees,
-                        parent_id,
-                        id,
-                    },
-                    resp_sender,
-                })
-                .expect("Did block processor thread panic?"),
-            None => resp_sender
-                .send(Err(format_err!("Executor is shutting down.")))
-                .expect("Failed to send error message."),
+        let _timer = OP_COUNTERS.timer("block_execute_time_s");
+        // Construct a StateView and pass the transactions to VM.
+        let state_view = VerifiedStateView::new(
+            Arc::clone(&self.storage_read_client),
+            committed_trees.version(),
+            committed_trees.state_root(),
+            parent_trees.state_tree(),
+        );
+
+        let vm_outputs = {
+            let _timer = OP_COUNTERS.timer("vm_execute_block_time_s");
+            V::execute_block(transactions.clone(), &self.vm_config, &state_view)?
+        };
+
+        let status: Vec<_> = vm_outputs
+            .iter()
+            .map(TransactionOutput::status)
+            .cloned()
+            .collect();
+        if !status.is_empty() {
+            debug!("Execution status: {:?}", status);
         }
-        resp_receiver
+
+        let (account_to_btree, account_to_proof) = state_view.into();
+        let output = Self::process_vm_outputs(
+            account_to_btree,
+            account_to_proof,
+            &transactions,
+            vm_outputs,
+            parent_trees,
+        )
+        .map_err(|err| format_err!("Failed to execute block: {}", err))?;
+
+        Ok(output)
     }
 
-    /// Commits a block and all its ancestors within a block batch. Returns `Ok(())` if successful.
+    /// Saves eligible blocks to persistent storage.
+    /// If we have multiple blocks and not all of them have signatures, we may send them to storage
+    /// in a few batches. For example, if we have
+    /// ```text
+    /// A <- B <- C <- D <- E
+    /// ```
+    /// and only `C` and `E` have signatures, we will send `A`, `B` and `C` in the first batch,
+    /// then `D` and `E` later in the another batch.
+    /// Commits a block and all its ancestors in a batch manner. Returns `Ok(())` if successful.
     pub fn commit_blocks(
         &self,
-        blocks: Vec<CommittableBlock>,
+        blocks: Vec<(Vec<Transaction>, Arc<ProcessedVMOutput>)>,
         ledger_info_with_sigs: LedgerInfoWithSignatures,
-    ) -> oneshot::Receiver<Result<()>> {
+        num_persistent_txns: u64,
```

Comment:
shouldn't we be able to derive this here? it seems like this is a bit fragile passing in an arbitrary integer on a public api.

Reply:
I changed it to `ExecutedTrees`